pipenv/core.py

```diff
@@ -1541,6 +1581,12 @@ def pip_install(
         pip_command.extend(["-r", vistir.path.normalize_path(r)])
     elif line:
         pip_command.extend(line)
+    if dev and use_constraint:
+        constraint_filename = write_constraint_to_file(
+            project,
+            requirements_dir=requirements_dir,
+        )
+        pip_command.extend(["-c", vistir.path.normalize_path(constraint_filename)])
```

Comment:
Is there an alternative function that would make sense other than calling something from vistir?  Also, is there a way to do this without writing to a file?

Reply:
Thanks for reviewing.

There are similar function to normalize path inside shell.py, I will replace vistir with this.
https://github.com/pypa/pipenv/blob/6c0c451b81fe08e4a11f275931b9952ec1182154/pipenv/utils/shell.py#L98-L101


After reading [constraints files](https://pip.pypa.io/en/stable/user_guide/#constraints-files), I think currently there are no way to pass constraint as argument to pip.

- The constraint file created from `write_to_constrain_file` is passed to subprocess, so it might be hard to not use the file.
https://github.com/pypa/pipenv/blob/30a6b1a5b82f0d04be14274b189467f10be2c3a2/pipenv/core.py#L1609-L1610

- The constraints files create from `default_constraint_file` inside resolver could be removed I think. But this might take time reading pip code base to see how **constraints files** in convert into **constraint line**.